### DISCUSSION TITLE
trouble with setValue on nested array path with useWatch changing from object to primitive

### DISCUSSION THREAD

**Discussion by ilovett** [react-hook-form/react-hook-form#3548] [Q&A]
I am running `setValue` in an `onChange` from a child component on a nested path like this:

```
   setValue("some.fieldArray[0].value", ""); // set the value to empty string -- previously was an object `{ number: 1, unit: 'day' }`
```

The useWatch value doesnt seem to update the first time changing from an object value to an empty string -- it seems to be delayed, the child useWatch doesnt seem to pick up the recently changed setValue until the second change...

Sorry no codesandbox I am getting sick of spending an hour recreating issues a codesandbox, I'm just looking for pointers or tips of why this might not be working as I expect... or workarounds...

Here is a gif of the ui interaction:

![2020-11-27 11 29 43](https://user-images.githubusercontent.com/1776100/100479879-6740c580-30a4-11eb-949c-731c3ea2fa38.gif)

**Comment by ilovett** [comment]
All the components are controlled, when the operator is "X Days Ago", then the Value component is a fragment with 3 hidden inputs and one text input which build out the object.... `value.amount`, `value.unit`, `value.direction` -- etc...

**Comment by ilovett** [comment]
I am thinking perhaps this only happens when changing a value from an object to a primitive...

```jsx
<input type="hidden" name="nested.path.value.one" value=1 />
<input type="hidden" name="nested.path.value.two" value=2 />
```

```
// nested.path.value === { one: 1, two: 2 }

setValue("nested.path.value", ""); // change object to primitive (empty string)

// useWatch still returns for `nested.path.value` === { one: 1, two: 2 } ....  (it should be returning "")
```

but now the DOM controller will only render:

```
<input type="text" name="nested.path.value" value="[object Object]" />
```

The next time `setValue("nested.path.value", "")` is called then the text input is correctly cleared...

**Reply by ilovett** [reply]
It seems like DOM is driving the returned `useWatch` value more than the `setValue` is...  I currently have a hack to setTimeout (wait for the DOM to update and no longer refer to object structure)....

**Reply by bluebill1049** [reply]
you can probably cheat on this without an actual input:

`<Controller name="nested.path.value.two" />`

**Comment by bluebill1049** [comment]
sorry for the late response, i have been busy not get in time to look at this. I will (if not ping me again)
